### PR TITLE
fix: Avoid deadlock in AgentReportStats Close during agent Close

### PR DESCRIPTION
**Pull request by mafredri**
Since AgentReportStats takes a stats function which was doing mutex
locking on agent shutdown, it was possible for there to be a deadlock
depending on how the AgentReportsStats Close function is implemented.

This mostly seems to happen on Windows test runners as it's pretty hard
to hit this edge case. The bug currently only exists in the test
implementation of AgentReportStats, however, this was refactored to be
more robust in case of future changes.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -226,6 +226,25 @@ func (a *agent) run(ctx context.Context) error {
...
+			a.logger.Error(ctx, "report stats", slog.Error(err))
+		} else {
+			if err = a.trackConnGoroutine(func() {
+				// This is OK because the agent never re-creates the tailnet
+				// and the only shutdown indicator is agent.Close().
+				<-a.closed
+				_ = cl.Close()
+			}); err != nil {
+				a.logger.Error(ctx, "report stats goroutine", slog.Error(err))
+				_ = cl.Close()
```
Comment (mafredri) [inline agent/agent.go:245]: Review: Is stats reporter startup failure fatal? I.e. do we return an error or not?

File: agent/agent.go [hunk 2]
```diff
@@ -561,28 +580,6 @@ func (a *agent) init(ctx context.Context) {
 	}
 
 	go a.runLoop(ctx)
-	cl, err := a.client.AgentReportStats(ctx, a.logger, func() *codersdk.AgentStats {
-		stats := map[netlogtype.Connection]netlogtype.Counts{}
-		a.closeMutex.Lock()
```
Comment (mafredri) [inline agent/agent.go:566]: Review: The deadlock happens between https://github.com/coder/coder/blob/b10368517082fd77fabcf9c0791e92af5d3766c5/agent/agent_test.go#L898 and https://github.com/coder/coder/blob/b10368517082fd77fabcf9c0791e92af5d3766c5/agent/agent_test.go#L909.

**Review by kylecarbs — APPROVED** [review]

**Review by mtojek — COMMENTED** [review]
The part that is confusing for me is that you're modifying the agent initialization phase, and you have to assume that the agent or its network is already dead. Maybe we should split it into a few elements?

File: agent/agent.go [hunk 3]
```diff
@@ -226,6 +226,25 @@ func (a *agent) run(ctx context.Context) error {
...
+
+		// Report statistics from the created network.
+		cl, err := a.client.AgentReportStats(ctx, a.logger, func() *codersdk.AgentStats {
+			stats := network.ExtractTrafficStats()
+			return convertAgentStats(stats)
+		})
+		if err != nil {
+			a.logger.Error(ctx, "report stats", slog.Error(err))
+		} else {
+			if err = a.trackConnGoroutine(func() {
```
[the remark this answers is not in the document]
Comment (mafredri) [reply agent/agent.go:238]: Do you mean the `logger.Error` below? The only error from `trackConnGoroutine` is agent closed, so yeah it's kinda pointless.

**Pull request merged**